content/posts/2021-06-25-Treasury-Expenditure-Bug.md

```diff
@@ -0,0 +1,167 @@
+
+## The Bootstrap Expenditure Amount
+
+Basing the maximum expenditure policy check on the average past _expenditure_ as opposed to the absolute past _income_ introduces an additional challenge: what to do at the start of the new treasury accounting system, when there are no past expenditures. Or, more generally, what the maximum allowed expenditure should be when there are no spends within the past window of blocks used as reference.
+
+Thus a new chain parameter was introduced: the expenditure bootstrap amount, which specifies how much DCR can be withdrawn from the treasury when the immediate history has no treasury spend transactions.
+
+For mainnet, this value was chosen to be 16k DCR, which was a value close to the top of the amount spent in the preceding months, maintaining the ability of the decentralized treasury to keep paying the existing contractors.
+
+It was the implementation of this specific feature that caused the policy check to fail when a test treasury spend transaction is present in the immediate history.
+
+## Triggering the Issue
+
+The code related to the bootstrap amount can be summarized as follows (this is an edited section of the [`maxTreasuryExpenditure()`](https://github.com/decred/dcrd/blob/afff2fdbcd4c57ade4f0d13e78ad2d3efaebcdec/blockchain/treasury.go#L650) function, redacted for clarity):
+
+```go
+	policyWindow := b.chainParams.TreasuryVoteInterval *
+		b.chainParams.TreasuryVoteIntervalMultiplier *
+		b.chainParams.TreasuryExpenditureWindow
+
+	// ...
+
+	// Next, sum up all tspends inside the N prior policy windows. If a
+	// given policy window does not have _any_ tspends, it isn't counted
+	// towards the average.	
+	for i := uint64(0); i < b.chainParams.TreasuryExpenditurePolicy && node != nil; i++ {
+		var spent int64
+		spent, _, node, err = b.sumPastTreasuryExpenditure(node, policyWindow)
+		// ...
+
+		if spent > 0 {
+			spentPriorWindows += spent
+			nbNonEmptyWindows++
+		}
+	}
+
+	// Calculate the average spent in each window. If there were _zero_
+	// prior windows with tspends, fall back to using the bootstrap
+	// average.
+	var avgSpentPriorWindows int64
+	if nbNonEmptyWindows > 0 {
+		avgSpentPriorWindows = spentPriorWindows / nbNonEmptyWindows
+	} else {
+		avgSpentPriorWindows = int64(b.chainParams.TreasuryExpenditureBootstrap)
+	}
+
+	//...
+```
+
+Note the last `if` statement: if there is at least one preceding window in which a treasury spend tx ("tspend") was mined, then the average used to derive the limit is calculated as the total spent across all tspends over the number of non-empty windows. Otherwise, the bootstrap amount is used.
```

Comment:
```suggestion
Note the last `if` statement: if there is at least one preceding window in which a treasury spend transaction ("tspend") was mined, then the average used to derive the limit is calculated as the total spent across all tspends over the number of non-empty windows. Otherwise, the bootstrap amount is used.
```